[Doc] Update Python test doc (#3011)
taichi-dev/taichi: docs/lang/articles/contribution/write_test.md

    pass
```

> Add some tests that makes use of `ti.log10` to ensure it works

Reviewer: ```suggestion
Add some tests that make use of `ti.log10` to ensure it works
```

import taichi as ti

# will test against both CPU and CUDA backends

> @ti.test(arch = [ti.cpu, ti.cuda])

Reviewer: ```suggestion
@ti.test(arch=[ti.cpu, ti.cuda])
```

    assert r[None] == 2
```

> In this case, python test will execute this function on `ti.cpu` and `ti.cuda` backend separately.

Reviewer: ```suggestion
In this case, Taichi will execute this test case on both `ti.cpu` and `ti.cuda` backends.
```

> And you may test against **all backends** by simply not specifying the

Reviewer 2: nit: ...all (available?) backends... or backends (depending on the system)... ;/

## Using `ti.approx` for comparison with tolerance

> Sometimes the precision of math operations could be relatively low on certain backends such as OpenGL,

Reviewer 2: nit: ...operations could be limited...? ;/

> e.g. `ti.log10(100)` may return `2.000001` or `1.999999` in this case.

Reviewer: ```suggestion
e.g., `ti.log10(100)` may return `2.000001` or `1.999999` in this case.
```

def test_sparse_field():
    # ... (some tests that requires sparse feature which is not supported by OpenGL)
```

## Requiring extension in tests

> If the test is depend on some extensions, you could add argument `require` in `@ti.test` decorator.

Reviewer: ```suggestion
If a test case depends on some extensions, you should add an argument `require` in the `@ti.test` decorator.
```

```python {1}
@ti.test(require=ti.extension.sparse)
def test_struct_for_pointer_block():
    n = 16
    block_size = 8

    f = ti.field(dtype=ti.f32)

    block = ti.root.pointer(ti.ijk, n // block_size)
    block.dense(ti.ijk, block_size).place(f)

    f[0, 2, 3] = 1

    @ti.kernel
    def count() -> int:
        tot = 0
        for I in ti.grouped(block):
            tot += 1
        return tot

    assert count() == 1
```

> Now, Taichi is supporting following extensions:

Reviewer: ```suggestion
Now, Taichi supports the following extensions:
```

> | Name          | Extension usage                                               |

Reviewer: ```suggestion
| Name          | Extension details                                             |
```

| ------------- | ------------------------------------------------------------- |
| sparse        | Sparse data structures                                        |
| async_mode    | Asynchronous execution mode                                   |
| quant_basic   | Basic operations in quantization                              |

> | quant         | Quantization                                                  |

Reviewer: ```suggestion
| quant         | Full quantization functionalities                             |
```

> | quant         | Quantization                                                  |

Reviewer: ```suggestion
| quant         | Basic quantization functionalities                            |
```

> | data64        | 64-bit data (Metal doesn't support 64-bit data buffers yet)   |

Reviewer: ```suggestion
| data64        | 64-bit data and arithmetics                                   |
```

> | adstack       | For keeping the history of mutable local variables            |

Reviewer: ```suggestion
| adstack       | For keeping the history of mutable local variables in autodiff|
```

| bls           | Block-local storage                                           |
| assertion     | Run-time asserts in Taichi kernels                            |

> | extfunc       | Invoke external functions or backend source                   |

Reviewer: ```suggestion
| extfunc       | Support inserting external functions calls or backend source  |
```

> | extfunc       | Support inserting external functions calls or backend source  |

Reviewer 2: ... external function calls or...

> | packed        | Packed mode, Shape will not be padded to a power of two       |

Reviewer: ```suggestion
| packed        | Packed mode: shapes will not be padded to powers of two       |
```

> | dynamic_index | Dynamic index support for both global and local tensors       |

Reviewer: ```suggestion
| dynamic_index | Dynamic index support for matrices and vectors                |
```
